File: Roadmap/19 - ENUMERACIONES/javascript/nlarrea.js
/*
 * EJERCICIO:
 * Empleando tu lenguaje, explora la definición del tipo de dato
 * que sirva para definir enumeraciones (Enum).
 * Crea un Enum que represente los días de la semana del lunes
 * al domingo, en ese orden. Con ese enumerado, crea una operación
 * que muestre el nombre del día de la semana dependiendo del número entero
 * utilizado (del 1 al 7).
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un pequeño sistema de gestión del estado de pedidos.
 * Implementa una clase que defina un pedido con las siguientes características:
 * - El pedido tiene un identificador y un estado.
 * - El estado es un Enum con estos valores: PENDIENTE, ENVIADO, ENTREGADO y CANCELADO.
 * - Implementa las funciones que sirvan para modificar el estado:
 *   - Pedido enviado
 *   - Pedido cancelado
 *   - Pedido entregado
 *   (Establece una lógica, por ejemplo, no se puede entregar si no se ha enviado, etc...)
 * - Implementa una función para mostrar un texto descriptivo según el estado actual.
 * - Crea diferentes pedidos y muestra cómo se interactúa con ellos. 
 */


// In JS there are no Enums -> we create one
class Enum {
    object = {};

    add(key, value) {
        this.object[key] = value;
    }

    getById(valueId) {
        const keys = Object.keys(this.object);
        if (typeof keys[valueId - 1] === 'undefined') {
            throw new Error('Value Not Found');
        }

        return keys[valueId - 1];
    }
}


class WeekDay extends Enum {
    constructor () {
        super();
    }

    object = {
        MONDAY: 1,
        TUESDAY: 2,
        WEDNESDAY: 3,
        THURSDAY: 4,
        FRIDAY: 5,
        SATURDAY: 6,
        SUNDAY: 7
    };
};
    

const weekdays = new WeekDay();
console.log(weekdays.getById(5));  // friday


/*
 * DIFICULTAD EXTRA (opcional):
 * Crea un pequeño sistema de gestión del estado de pedidos.
 * Implementa una clase que defina un pedido con las siguientes características:
 * - El pedido tiene un identificador y un estado.
 * - El estado es un Enum con estos valores: PENDIENTE, ENVIADO, ENTREGADO y CANCELADO.
 * - Implementa las funciones que sirvan para modificar el estado:
 *   - Pedido enviado
 *   - Pedido cancelado
 *   - Pedido entregado
 *   (Establece una lógica, por ejemplo, no se puede entregar si no se ha enviado, etc...)
 * - Implementa una función para mostrar un texto descriptivo según el estado actual.
 * - Crea diferentes pedidos y muestra cómo se interactúa con ellos. 
 */


class OrderStatus extends Enum {
    constructor () {
        super();
    }

    object = {
        PENDING: 1,
        SHIPPED: 2,
        DELIVERED: 3,
        CANCELED: 4
    }
};


class Order {
    constructor (orderId) {
        this.orderId = orderId;
        this.state = new OrderStatus().getById(1);

        console.log(`\nThe order '${this.orderId}' is ${this.state}`);
    };

    updateState() {
        if (this.state == 'CANCELED' || this.state == 'DELIVERED') {
            // If order is canceled or delivered -> maintain the state as it is
            return;
        } else if (this.state == 'PENDING') {
            this.state = new OrderStatus().getById(2);
        } else if (this.state == 'SHIPPED') {
            this.state = new OrderStatus().getById(3);
        }
    }

    cancelOrder() {
        if (this.state != 'DELIVERED') {
            this.state = new OrderStatus().getById(4);
        }
    }

    printCurrentState() {
        console.log(`The order '${this.orderId}' is ${this.state}`);
    }
}

const order1 = new Order('1234');  // The order '1234' is PENDING
order1.updateState();
order1.printCurrentState();  // shipped
order1.updateState();
order1.printCurrentState();  // delivered
order1.cancelOrder();  // it won't work because it's delivered already
order1.printCurrentState();  // delivered

const order2 = new Order("5678");  // pending
order2.updateState();
order2.printCurrentState();  // shipped
order2.cancelOrder();
order2.printCurrentState();  // canceled
order2.updateState();  // it won't work because it's canceled
order2.printCurrentState();  // canceled

const order3 = new Order("1290");  // pending
order3.cancelOrder();
order3.printCurrentState();  // canceled

File: Roadmap/19 - ENUMERACIONES/python/nlarrea.py
"""
 * EJERCICIO:
 * Empleando tu lenguaje, explora la definición del tipo de dato
 * que sirva para definir enumeraciones (Enum).
 * Crea un Enum que represente los días de la semana del lunes
 * al domingo, en ese orden. Con ese enumerado, crea una operación
 * que muestre el nombre del día de la semana dependiendo del número entero
 * utilizado (del 1 al 7).
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un pequeño sistema de gestión del estado de pedidos.
 * Implementa una clase que defina un pedido con las siguientes características:
 * - El pedido tiene un identificador y un estado.
 * - El estado es un Enum con estos valores: PENDIENTE, ENVIADO, ENTREGADO y CANCELADO.
 * - Implementa las funciones que sirvan para modificar el estado:
 *   - Pedido enviado
 *   - Pedido cancelado
 *   - Pedido entregado
 *   (Establece una lógica, por ejemplo, no se puede entregar si no se ha enviado, etc...)
 * - Implementa una función para mostrar un texto descriptivo según el estado actual.
 * - Crea diferentes pedidos y muestra cómo se interactúa con ellos. 
"""

from enum import Enum


class WeekDay(Enum):
    MONDAY = 1
    TUESDAY = 2
    WEDNESDAY = 3
    THURSDAY = 4
    FRIDAY = 5
    SATURDAY = 6
    SUNDAY = 7


def print_week_day(day_index: int) -> None:
    if day_index < 1 or day_index > 7:
        print("You should enter a number between 1 and 7 (both included).")
        return

    print(WeekDay(day_index).name.capitalize())


print_week_day(5)  # Friday


"""
 * DIFICULTAD EXTRA (opcional):
 * Crea un pequeño sistema de gestión del estado de pedidos.
 * Implementa una clase que defina un pedido con las siguientes características:
 * - El pedido tiene un identificador y un estado.
 * - El estado es un Enum con estos valores: PENDIENTE, ENVIADO, ENTREGADO y CANCELADO.
 * - Implementa las funciones que sirvan para modificar el estado:
 *   - Pedido enviado
 *   - Pedido cancelado
 *   - Pedido entregado
 *   (Establece una lógica, por ejemplo, no se puede entregar si no se ha enviado, etc...)
 * - Implementa una función para mostrar un texto descriptivo según el estado actual.
 * - Crea diferentes pedidos y muestra cómo se interactúa con ellos. 
"""


class OrderStatus(Enum):
    PENDING = 1
    SHIPPED = 2
    DELIVERED = 3
    CANCELED = 4


class Order:
    def __init__(self, order_id: str) -> None:
        self.order_id = order_id
        self.state = OrderStatus.PENDING

        print(f"\nThe order '{self.order_id}' is {self.state.name}")

    def update_state(self):
        if (
            self.state == OrderStatus.CANCELED
            or self.state == OrderStatus.DELIVERED
        ):
            # If order is canceled or delivered -> maintain the state as it is
            return

        self.state = OrderStatus(self.state.value + 1)

    def cancel_order(self):
        if self.state != OrderStatus.DELIVERED:
            self.state = OrderStatus.CANCELED

    def print_current_state(self):
        print(f"The order '{self.order_id}' is {self.state.name}")


order_1 = Order("1234")  # pending
order_1.update_state()
order_1.print_current_state()  # shipped
order_1.update_state()
order_1.print_current_state()  # delivered
order_1.cancel_order()  # it won't work because it's delivered already
order_1.print_current_state()  # delivered

order_2 = Order("5678")  # pending
order_2.update_state()
order_2.print_current_state()  # shipped
order_2.cancel_order()
order_2.print_current_state()  # canceled
order_2.update_state()  # it won't work because it's canceled
order_2.print_current_state()  # canceled

order_3 = Order("1290")  # pending
order_3.cancel_order()
order_3.print_current_state()  # canceled
